lib/uptime/env.ex

```diff
@@ -27,23 +29,82 @@ defmodule Uptime.Env do
     :ok
   end
 
+  @doc """
+  Read a value from the environment.
+  First check for a secrets file, then the environment, then the config file.
+  secrets file > environment variable > config file
+  """
   @spec read(String.t()) :: String.t()
   def read(name) do
-    case File.read("/etc/secrets/#{name}") do
-      {:ok, value} -> value
-      _not_found -> System.get_env(name)
+    prefixed_name = "#{@prefix}_#{name}"
+
+    file_contents =
+      :supported_secrets_dir
+      |> get("/etc/secrets")
+      |> Path.join(prefixed_name)
+      |> File.read()
+
+    case file_contents do
+      {:ok, value} ->
+        value
+
+      _not_found ->
+        case read_from_env(prefixed_name) do
+          {:ok, value} ->
+            value
+
+          _not_found ->
+            name
+            |> String.downcase()
+            |> read_from_config_file(find_config_path())
+        end
+    end
+  end
+
+  @spec read_from_env(String.t()) :: {:ok, String.t()} | :error
+  defp read_from_env(name) do
+    value = System.get_env(name)
+
+    if valid_env_var_value(value) do
+      {:ok, value}
+    else
+      :error
+    end
+  end
+
+  @spec read_from_config_file(String.t(), String.t()) :: any() | nil
+  defp read_from_config_file(key, path) when is_binary(path) do
+    case YamlElixir.read_from_file(path) do
+      {:ok, config} ->
+        # TODO: Implement a deep get function when we add nested configuration structure
+        config[key]
+
+      _not_found ->
+        nil
     end
   end
 
+  defp read_from_config_file(_key, path) when is_nil(path), do: nil
+
+  @spec find_config_path() :: String.t() | nil
+  defp find_config_path do
+    @supported_config_filenames
+    |> Enum.reduce([], fn file, acc ->
+      acc ++
+        Enum.map(get(:supported_config_dirs, ["/etc/uptimes/config.yaml"]), &Path.join(&1, file))
+    end)
+    |> Enum.find(&File.exists?/1)
+  end
+
   @spec has?(String.t()) :: boolean()
   def has?(name) do
     name
     |> read()
     |> valid_env_var_value()
   end
 
-  @spec get(atom()) :: any()
-  def get(key), do: Application.get_env(:uptime, key)
+  @spec get(atom(), any()) :: any()
+  def get(key, default \\ nil), do: Application.get_env(:uptime, key, default)
```

Comment:
While I think normally it can be nice to run env vars into runtime.exs first, I think it may be different with the user defined env vars (e.g. BASIC_AUTH_USERNAME).

For example, if we just use `read/1` on the fly from env.ex then users could update the config and the app would read the new values. Also what about config dependencies from app in env.ex (I’m adding supported_config_dirs: [] so we could require this option in config and also use env.ex in config, right?)

Reply:
so generally just following on security best practices you don't want to let the current runtime be manipulated like that, if you change the configuration value the docker image should be restarted, also in the case of secrets that would take a hit on reading from the file system each time, which is the benefit of putting in the Application env as it is now stored in memory

Reply:
Ok that sounds good